internal/scheduler/context/context.go

```diff
@@ -699,33 +681,95 @@ func (jctx *JobSchedulingContext) GetNodeSelector(key string) (string, bool) {
+func EmptyGangInfo(job interfaces.LegacySchedulerJob) GangInfo {
+	return GangInfo{
+		// An Id of "" indicates that this job is not in a gang; we set
+		// Cardinality and MinimumCardinality (as well as the other fields,
+		// which all make sense in this context) accordingly.
+		Id:                 "",
+		Cardinality:        1,
+		MinimumCardinality: 1,
+		PriorityClassName:  job.GetPriorityClassName(),
+		NodeUniformity:     job.GetAnnotations()[configuration.GangNodeUniformityLabelAnnotation],
+	}
+}
+
+func GangInfoFromLegacySchedulerJob(job interfaces.LegacySchedulerJob) (GangInfo, error) {
+	gangInfo := EmptyGangInfo(job)
+
+	annotations := job.GetAnnotations()
+
+	gangId, ok := annotations[configuration.GangIdAnnotation]
+	if !ok {
+		return gangInfo, nil
+	}
+	if gangId == "" {
+		return gangInfo, errors.Errorf("gang id is empty")
+	}
+
+	gangCardinalityString, ok := annotations[configuration.GangCardinalityAnnotation]
+	if !ok {
+		return gangInfo, errors.Errorf("annotation %s is missing", configuration.GangCardinalityAnnotation)
+	}
+	gangCardinality, err := strconv.Atoi(gangCardinalityString)
+	if err != nil {
+		return gangInfo, errors.WithStack(err)
+	}
+	if gangCardinality <= 0 {
+		return gangInfo, errors.Errorf("gang cardinality %d is non-positive", gangCardinality)
+	}
+
+	gangMinimumCardinalityString, ok := annotations[configuration.GangMinimumCardinalityAnnotation]
+	if !ok {
+		// If it is not set, use gangCardinality as the minimum gang size.
+		gangMinimumCardinalityString = gangCardinalityString
+	}
+	gangMinimumCardinality, err := strconv.Atoi(gangMinimumCardinalityString)
+	if err != nil {
+		return gangInfo, errors.WithStack(err)
+	}
+	if gangMinimumCardinality <= 0 {
+		return gangInfo, errors.Errorf("gang minimum cardinality %d is non-positive", gangMinimumCardinality)
+	}
+	if gangMinimumCardinality > gangCardinality {
+		return gangInfo, errors.Errorf("gang minimum cardinality %d is greater than gang cardinality %d", gangMinimumCardinality, gangCardinality)
+	}
+
+	gangInfo.Id = gangId
+	gangInfo.Cardinality = gangCardinality
+	gangInfo.MinimumCardinality = gangMinimumCardinality
+	return gangInfo, nil
+}
+
+func JobSchedulingContextsFromJobs[J interfaces.LegacySchedulerJob](priorityClasses map[string]types.PriorityClass, jobs []J) []*JobSchedulingContext {
 	jctxs := make([]*JobSchedulingContext, len(jobs))
 	for i, job := range jobs {
-		jctxs[i] = JobSchedulingContextFromJob(priorityClasses, job, extractGangInfo)
+		jctxs[i] = JobSchedulingContextFromJob(priorityClasses, job)
 	}
 	return jctxs
 }
 
-func JobSchedulingContextFromJob(priorityClasses map[string]types.PriorityClass, job interfaces.LegacySchedulerJob, extractGangInfo func(map[string]string) (string, int, int, bool, error)) *JobSchedulingContext {
-	// TODO: Move cardinality to gang context only and remove from here.
-	// Requires re-phrasing nodedb in terms of gang context, as well as feeding the value extracted from the annotations downstream.
```

Comment:
I've deleted this comment, because I think that it's misleading: the scope of the required changes is much larger (for example, this would also require fundamental changes to the in-memory job database).